nixos/modules/config/nix-channel.nix

```diff
@@ -49,22 +79,30 @@ in
   config = mkIf cfg.enable {
 
     environment.extraInit =
-      ''
+      mkIf cfg.channel.enable ''
         if [ -e "$HOME/.nix-defexpr/channels" ]; then
           export NIX_PATH="$HOME/.nix-defexpr/channels''${NIX_PATH:+:$NIX_PATH}"
         fi
       '';
 
+    environment.extraSetup = mkIf (!cfg.channel.enable) ''
+      rm $out/bin/nix-channel
+    '';
+
+    # NIX_PATH has a non-empty default according to Nix docs, so we don't unset
+    # it when empty.
     environment.sessionVariables = {
       NIX_PATH = cfg.nixPath;
     };
 
-    system.activationScripts.nix-channel = stringAfter [ "etc" "users" ]
-      ''
+    nix.settings.nix-path = mkIf (! cfg.channel.enable) (mkDefault "");
```

Comment:
I don't think we want to set this to empty string here because that clears the default nix-path setting of nix which breaks many programs like nix-tree or comma which rely on it to find some nixpkgs which in my case points to a flake.

Reply:
I don't think so. The point of this is one really *doesn't want* to use a populated `NIX_PATH`. If other programs can't cope with that, they should be fixed.

One can always set `nix.settings.nix-path` elsewhere than this module instead.

Reply:
By default nix-path is left empty which means it default to NIX_PATH env. When we explicitly set it to empty, that is no longer the case and things completely break for many tools which makes this option rather useless.


```
 ➜ nix show-config | rg nix-path
nix-path = nixpkgs=/nix/store/x6ly3m0dylzsbrgz0sx937xxcyswr2yp-source nixos=/nix/store/x6ly3m0dylzsbrgz0sx937xxcyswr2yp-source nixos-config=/etc/nixos/configuration.nix

 ➜ (unset NIX_PATH; nix show-config | rg nix-path)
nix-path =

 ➜ (unset NIX_PATH; nix --option nix-path "" show-config | rg nix-path)
nix-path =

 ➜ nix --option nix-path "" show-config | rg nix-path
nix-path =
```

Reply:
That seems like the behaviour that someone ripping out a formerly-mandatory system component to avoid non-flakes things would want, given that properly flakes-aware tools like `nixos-rebuild(8)` won't look at the Nix path when used correctly.

Reply:
hmmm, yeah, that sounds about right. Maybe I misunderstood the feature a bit.

Reply:
> I don't think so. The point of this is one really _doesn't want_ to use a populated `NIX_PATH`

Kind of debatable, IMO. Feels like you're assuming that disabling channels automatically implies that:

- The person is using flakes
- The person wants to use a PURE flake workflow everywhere.

The reality is that:
- nixpkgs is kind of special
- You can use --impure with flakes
- Some people may want a more "it just works" sort of thing with certain types of commands/ and /or the ability to build their system against a flake and then have all other commands use the same version of nixpkgs that their system was built against, or maybe still allow the NIX_PATH variable to take effect.

Reply:
> That seems like the behaviour that someone ripping out a formerly-mandatory system component to avoid non-flakes things would want, given that properly flakes-aware tools like `nixos-rebuild(8)` won't look at the Nix path when used correctly.

This is a weird position that doesn't make sense.

You can only use NIX_PATH at all, if you pass the --impure flag to these commands. If you dont, you'll get:

```
error: cannot look up '<nixpkgs>' in pure evaluation mode (use '--impure' to override)
```

If you're explicitly opting IN by passing the --impure flag, this behavior makes absolutely no sense.

Really, the problem is not setting nix-path to the empty string, the problem is https://github.com/NixOS/nix/issues/8890 and https://github.com/NixOS/nix/issues/8890

Reply:
I'm also encountered the problem where I don't use channels but put aliases in NIX_PATH for commands. This setting disables NIX_PATH envvar entirely, which is not what I want. I'm kinda thinking we should revert this line.